### PR TITLE
[core-data] throwOnError option for saveEntityRecord and deleteEntityRecord actions

### PR DESCRIPTION
**Pull request by adamziel**
## Description

This is a continuation of [Entity records hooks](https://github.com/WordPress/gutenberg/pull/38135).

An original idea was to make creating, editing, and deleting entity records easier by introducing a specialized convenience wrapper such as `useEntityMutation`:

```js
const {
	edit,
	editedRecord,
	hasEdits,
	saveEdits,
	isSaving,
	saveError
} = useEntityMutation( 'postType', 'page', pageId );
```

The rationale – that the full flow is not easy to figure out, confusing at times, and it is easy to omit a part or two. For example,  to edit a record, save it, and display a progress indicator and any error information, you need at least the following information and actions:

```js
const {
	editEntityRecord,
	saveEditedEntityRecord
} = useDispatch( coreStore );

const { editedRecord, hasEdits, isSaving, saveError } = useSelect(
	( select ) => {
		const args = [kind, type, id];
		return {
			editedRecord: select( coreStore ).getEditedEntityRecord( ...args ),
			hasEdits: select( coreStore ).hasEditsForEntityRecord( ...args ),
			isSaving: select( coreStore ).isSavingEntityRecord( ...args ),
			saveError: select( coreStore ).getLastEntitySaveError( ...args ),
		};
	},
	[kind, type, id],
);
```

Diving deeper, we realized that it is hard to distinguish between creating and updating records when both use-cases are crammed within a single `useEntityMutation` hook.

In this PR, I explored three smaller hooks called `useEntityRecordCreate`, `useEntityRecordUpdate`, and `useEntityRecordDelete`. See the [relevant commit](https://github.com/WordPress/gutenberg/pull/39258/commits/fb561cf6fe320c307eafac61e13fd1091a7b56cb).

However, it quickly occurred to me that we only rely on all these selectors because we do not embrace the full power of promises. For example, the Gutenberg codebase commonly relies on the following pattern:

```js
	const templatePart = await saveEntityRecord(
		'postType',
		'wp_template_part',
		{
			slug: cleanSlug,
			title,
			content: '',
			area,
		}
	);

	const lastEntitySaveError = getLastEntitySaveError(
		'postType',
		'wp_template_part',
		templatePart.id
	);
	if ( lastEntitySaveError ) {
		throw lastEntitySaveError;
	}
```

This looks a whole lot like what we do in PHP:

```php
$result = create_post();
if ( is_wp_error( $result ) ) {
    return $result;
}
```

What if we used promises instead? Even better, what if we ditched the hooks and used the promises right in the store so that other stores could benefit from that as well? And so this is the proposal this PR makes:

```js
try {
	const templatePart = await saveEntityRecord(
		'postType',
		'wp_template_part',
		{
			slug: cleanSlug,
			title,
			content: '',
			area,
		},
		{ throwOnError: true }
	);
} catch ( e ) {
	// Display an error notice
}
```

It takes care of:

* The errors, since they're thrown.
* The loading state, since we know whether the promise is still running or not. Of course, `isSavingEntityRecord` may still be used when needed or more convenient.

## Different angles

Re-throwing errors is different than having a convenience wrapper for actions like `editEntityRecord`, `saveEditedEntityRecord`, `saveEntityRecord` and the `getEditedEntityRecord` selector. Perhaps that's where a hook would actually shine:

```js
const { create, isCreating, error } = useEntityRecordCreate();
const { editedRecord, applyEdits, saveEdits, isSaving, error } = useEntityRecordEdit();
const { remove, isRemoving, error } = useEntityRecordRemove(); // it's "remove", because "delete" is a reserved keyword
```

However, given the "throwing" actions, accessing `error` and `isCreating/isSaving/isRemoving` would not be needed as often, which leaves us with just:

```js
const { create } = useEntityRecordCreate();
const { editedRecord, applyEdits, saveEdits } = useEntityRecordEdit();
const { remove } = useEntityRecordRemove();
```

Which doesn't do much for creating and removing records, so we could also ditch these hooks and settle only for the `useEntityRecordEdit` hook:

```js
const { editedRecord, applyEdits, saveEdits } = useEntityRecordEdit();
```

I'm interested in your thoughts about all that.

cc @gziolo @kevin940726 @getdave @noisysocks @scruffian @draganescu

### CODE REVIEW

**Comment by draganescu** [comment]
Looking at the resulting code, somehow [the more verbose code](https://github.com/WordPress/gutenberg/pull/39258/files#diff-bf9863e9cdc2891132a3751d051b5171d1ec7d4856491870f1f941e756fedcfcL290) seems more readable? It's like a story.

**Comment by noisysocks** [comment]
You didn't used to be able to `await` an action like you can now which is why the API is a bit clunky 🙂

I like the idea, but I don't find the API to be very intuitive if I'm honest. That is, when I read the code, I don't instantly know what it will do. What do you think about a param?

```js
await saveEditedEntityRecord(
 	'postType',
 	template.type,
 	template.id,
	{ throwOnError: true }
);
```

File: packages/core-data/src/actions.js [hunk 1]
```diff
@@ -759,3 +759,69 @@ export function receiveAutosaves( postId, autosaves ) {
...
+	const [ kind, name, id ] = args;
+	const error = select.getLastEntityDeleteError( kind, name, id );
+
+	if ( error || ! result ) {
+		throw error;
+	}
+	return result;
+};
+
+const getRecordPk = ( kind, name, record ) => async ( { dispatch } ) => {
```
Comment (gziolo) [inline packages/core-data/src/actions.js]: Nit: What does `Pk` stand for in the name of function?
```suggestion
const getRecordPrimaryKey = ( kind, name, record ) => async ( { dispatch } ) => {
```

Let's make it explicit so readers don't have to guess.

**Comment by gziolo** [comment]
I'm not familiar with the existing codebase that uses the flow that is discussed in this PR, but I must admit that the proposal from @noisysocks in https://github.com/WordPress/gutenberg/pull/39258#issuecomment-1064760211 sounds reasonable. This way we can reduce the boilerplate code and we still leave the choice to the developer how it should be handled. The challenge for public APIs is that you rather want to minimize the number of methods exposed to optimize for the most common usage and easier discovery. This goes in contrast to the consuming projects where a new method might help to make the code more readable. 

> This is a continuation of [Entity records hooks](https://github.com/WordPress/gutenberg/pull/38135).
> An original idea was to make creating, editing, and deleting entity records easier by introducing a specialized convenience wrapper such as `useEntityMutation:`

Thank you for sharing the full context and your thoughts process. It really helps to understand how you arrived at the final proposal and more importantly leaves a great archive that we can use in the future when evolving all those APIs around entities.

File: packages/core-data/src/actions.js (hunk 1 above)
Comment (adamziel) [reply packages/core-data/src/actions.js]: Good idea! Although we no longer need this function after applying changes proposed by @noisysocks :D

**Comment by adamziel** [comment]
> I like the idea, but I don't find the API to be very intuitive if I'm honest. That is, when I read the code, I don't instantly know what it will do. What do you think about a param?

I like it much better @noisysocks, thank you! I just updated this PR accordingly.

> Thank you for sharing the full context and your thoughts process. It really helps to understand how you arrived at the final proposal and more importantly leaves a great archive that we can use in the future when evolving all those APIs around entities.

My pleasure @gziolo ! I suspected that sharing just the final conclusion wouldn't be quite enough to explain the rationale here, I'm glad that was useful! I might start doing it more often when reasonable 🤔

File: packages/core-data/src/actions.js [hunk 2]
```diff
@@ -209,22 +209,26 @@ export function receiveEmbedPreview( url, preview ) {
...
+ * @param {string}   recordId                     Record ID of the deleted entity.
+ * @param {?Object}  query                        Special query parameters for the
+ *                                                DELETE API call.
+ * @param {Object}   [options]                    Delete options.
+ * @param {Function} [options.__unstableFetch]    Internal use only. Function to
+ *                                                call instead of `apiFetch()`.
+ *                                                Must return a promise.
+ * @param {boolean}  [options.throwOnError=false] Whether to re-throw exceptions.
+ *                                                If false, dispatching this action
+ *                                                always resolves successfully and.
```
Comment (gziolo) [inline packages/core-data/src/actions.js]: Nit: small thing that leaks into the README file 😅 
```suggestion
 *                                                always resolves successfully and
```

File: packages/core-data/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+–   The saveEntityRecord, saveEditedEntityRecord, and deleteEntityRecord actions now accept an optional throwOnError option (defaults to false). When set to true, any exceptions occurring when the action was executing are re-thrown, causing dispatch().saveEntityRecord() to reject with an error. ([#39258](https://github.com/WordPress/gutenberg/pull/39258))
```
Comment (gziolo) [inline packages/core-data/CHANGELOG.md:6]: We need a section header so the tool processing the changelog can match it and enforce a minor version upgrade:
```suggestion
### New Features

–   The saveEntityRecord, saveEditedEntityRecord, and deleteEntityRecord actions now accept an optional throwOnError option (defaults to false). When set to true, any exceptions occurring when the action was executing are re-thrown, causing dispatch().saveEntityRecord() to reject with an error. ([#39258](https://github.com/WordPress/gutenberg/pull/39258))
```

See https://github.com/WordPress/gutenberg/tree/trunk/packages#maintaining-changelogs for more details.

File: packages/core-data/src/actions.js [hunk 4]
```diff
@@ -386,20 +394,28 @@ export function __unstableCreateUndoLevel() {
...
+ * @param {string}   name                         Name of the received entity.
+ * @param {Object}   record                       Record to be saved.
+ * @param {Object}   options                      Saving options.
+ * @param {boolean}  [options.isAutosave=false]   Whether this is an autosave.
+ * @param {Function} [options.__unstableFetch]    Internal use only. Function to
+ *                                                call instead of `apiFetch()`.
+ *                                                Must return a promise.
+ * @param {boolean}  [options.throwOnError=false] Whether to re-throw exceptions.
+ *                                                If false, dispatching this action
+ *                                                always resolves successfully and.
```
Comment (gziolo) [inline packages/core-data/src/actions.js]: Nit: same thing 😄 
```suggestion
 *                                                always resolves successfully and
```

File: packages/edit-widgets/src/store/actions.js [hunk 5]
```diff
@@ -282,18 +282,13 @@ export const saveWidgetArea = ( widgetAreaId ) => async ( {
...
 	registry
 		.dispatch( coreStore )
-		.saveEditedEntityRecord( KIND, WIDGET_AREA_ENTITY_TYPE, widgetAreaId );
-	const saveErrorAfter = registry
-		.select( coreStore )
-		.getLastEntitySaveError( KIND, WIDGET_AREA_ENTITY_TYPE, widgetAreaId );
-	if ( saveErrorAfter && saveErrorBefore !== saveErrorAfter ) {
-		throw new Error( saveErrorAfter );
-	}
+		.throwingSaveEditedEntityRecord(
```
Comment (gziolo) [inline packages/edit-widgets/src/store/actions.js]: I guess you are still working on updating the usage.

**Comment by gziolo** [comment]
> Which doesn't do much for creating and removing records, so we could also ditch these hooks and settle only for the useEntityRecordEdit hook:
>
> ```js
> const { editedRecord, applyEdits, saveEdits } = useEntityRecordEdit();
> ```

We can also expose `applyEdits` and `saveEdits` methods from the existing `useEntityRecord`. I bet `editedRecord` is already there.

I guess for `create` and `delete` dispatching actions as presented in this PR in most cases is enough.

File: packages/edit-widgets/src/store/actions.js (hunk 5 above)
Comment (adamziel) [reply packages/edit-widgets/src/store/actions.js]: yes

File: packages/core-data/CHANGELOG.md (hunk 3 above)
Comment (adamziel) [reply packages/core-data/CHANGELOG.md:6]: Thank you!

File: packages/core-data/src/actions.js (hunk 4 above)
Comment (adamziel) [reply packages/core-data/src/actions.js]: good spot! I made it much shorter than that – how is it now?

File: packages/edit-widgets/src/store/actions.js (hunk 5 above)
Comment (adamziel) [reply packages/edit-widgets/src/store/actions.js]: finished

File: packages/core-data/src/actions.js (hunk 2 above)
Comment (adamziel) [reply packages/core-data/src/actions.js]: Let's discuss under your other note :-)

**Comment by adamziel** [comment]
The e2e failures are unrelated to this PR, see https://core.trac.wordpress.org/ticket/55367

File: docs/reference-guides/data/data-core.md [hunk 6]
```diff
@@ -536,6 +536,7 @@ _Parameters_
 -   _query_ `?Object`: Special query parameters for the DELETE API call.
 -   _options_ `[Object]`: Delete options.
 -   _options.\_\_unstableFetch_ `[Function]`: Internal use only. Function to call instead of `apiFetch()`. Must return a promise.
+-   _options.throwOnError_ `[boolean]`: If false, this action suppresses all the exceptions.
```
Comment (gziolo) [inline docs/reference-guides/data/data-core.md]: Let's clarify what's the default behavior here and in the other method.

File: packages/core-data/README.md [hunk 7]
```diff
@@ -70,6 +70,7 @@ _Parameters_
 -   _query_ `?Object`: Special query parameters for the DELETE API call.
 -   _options_ `[Object]`: Delete options.
 -   _options.\_\_unstableFetch_ `[Function]`: Internal use only. Function to call instead of `apiFetch()`. Must return a promise.
+-   _options.throwOnError_ `[boolean]`: If false, this action suppresses all the exceptions.
```
Comment (gziolo) [inline packages/core-data/README.md]: Aside: I think @michalczaplinski raised it in other places that it's confusing that the same APIs are documented in two places - here and in [docs/reference-guides/data/data-core.md](https://github.com/WordPress/gutenberg/pull/39258/files#diff-2b04a3ba2634ce97fb60b8d7765c171a769f4aaa3d39830b41830e4192018fba). 

It'd be a good follow-up.

**Review by gziolo — APPROVED** [review]
I'm personally happy about the current proposal so I mark this PR as approved. However, I would be more than happy to see feedback from other folks to ensure we are all on the same page.

File: packages/core-data/README.md (hunk 7 above)
Comment (adamziel) [reply packages/core-data/README.md]: The fix is beyond this PR, but I fully agree

Comment (gziolo) [reply packages/core-data/README.md]: Follow-up = new PR. So we are on the same page 😄

File: packages/core-data/src/actions.js [hunk 8]
```diff
@@ -273,6 +275,10 @@ export const deleteEntityRecord = (
 			error,
 		} );
 
+		if ( error && throwOnError ) {
```
Comment (kevin940726) [inline packages/core-data/src/actions.js]: What if `error` is `undefined` or `null` or `0` or any falsy value? Even though it's an edge case, it should still throw such error 😅. I guess we'll need a flag like `hasError` to determine if it reaches the `catch` clause or not.
(Would be awesome if there's a test case for it too, but it's a nitpick.)

**Review by kevin940726 — COMMENTED** [review]
Mostly looks good! Just a small concern regarding falsy error values.

**Review by kevin940726 — APPROVED** [review]
LGTM 👍

**Pull request merged**